framework/src/base/MooseApp.C

```diff
@@ -1573,128 +1573,174 @@ MooseApp::getMeshGeneratorOutput(const std::string & name)
+        // final generator we just pulled.
+        const auto & allValues = resolver.getSortedValues();
+        decltype(ancestors) all(allValues.begin(), allValues.end());
+
+        decltype(ancestors) ind_tree;
+        std::set_difference(all.begin(),
+                            all.end(),
+                            ancestors.begin(),
+                            ancestors.end(),
+                            std::inserter(ind_tree, ind_tree.end()));
+
+        std::ostringstream oss;
+        oss << "Your MeshGenerator tree contains multiple possible generator outputs :\n\""
+            << _final_generator_name
+            << " and one or more of the following from an independent set: \"";
+        bool first = true;
+        for (const auto & gen : ind_tree)
         {
-          // Need to remove duplicates and possibly perform a difference so we'll import out list
-          // into a set for these operations.
-          std::set<std::shared_ptr<MeshGenerator>> ancestors(ancestor_list.begin(),
-                                                             ancestor_list.end());
-          // Get all of the items from the resolver so we can compare against the tree from the
-          // final generator we just pulled.
-          const auto & allValues = resolver.getSortedValues();
-          decltype(ancestors) all(allValues.begin(), allValues.end());
-
-          decltype(ancestors) ind_tree;
-          std::set_difference(all.begin(),
-                              all.end(),
-                              ancestors.begin(),
-                              ancestors.end(),
-                              std::inserter(ind_tree, ind_tree.end()));
-
-          std::ostringstream oss;
-          oss << "Your MeshGenerator tree contains multiple possible generator outputs :\n\""
-              << _final_generator_name
-              << " and one or more of the following from an independent set: \"";
-          bool first = true;
-          for (const auto & gen : ind_tree)
-          {
-            if (!first)
-              oss << ", ";
-            else
-              first = false;
+          if (!first)
+            oss << ", ";
+          else
+            first = false;
 
-            oss << gen->name();
-          }
-          oss << "\"\n\nThis may be due to a missing dependency or may be intentional. Please "
-                 "select the final MeshGenerator in\nthe [Mesh] block with the \"final_generator\" "
-                 "parameter or add additional dependencies to remove the ambiguity.";
-          mooseError(oss.str());
+          oss << gen->name();
         }
+        oss << "\"\n\nThis may be due to a missing dependency or may be intentional. Please "
+               "select the final MeshGenerator in\nthe [Mesh] block with the \"final_generator\" "
+               "parameter or add additional dependencies to remove the ambiguity.";
+        mooseError(oss.str());
       }
+    }
+  }
+}
```

Comment:
I initially tried to delete `_mesh_generators` here after `_ordered_generators` is created. But got errors. So I end it up putting the clear at Line 1758.

Reply:
Shared pointers are easy to copy. You just do normal assignment and you have a copy. As long as somebody is holding a pointer to the underlying object, it will not be deleted. So I don't see how a clear of one vector would cause errors unless it holds the only pointer to an object. What was the error?

Reply:
I think doing `_mesh_generators.clear()` after `_ordered_generators` is created seems deleted all mesh generators. All mesh generator tests failed. Once I moved the clear later after all mesh generators are executed, every test works fine. I did not look into it further. But seems inside DependencyResolver, shared pointer copying is not doing the right thing. I have no much experience with shared pointer though.

Reply:
Btw, the diff seems a lot, but I actually did not change much in the code.

Reply:
The `_mesh_generators` vector is the "warehouse" for the generators. We don't have a second copy of those anywhere. They must not be cleared before execution.

Reply:
The issue might be that _mesh_generators are accessed through public APIs. I bet there are some assumptions about getting access to the generators externally, which are probably where your errors come from.